diff --git a/src/backend/Makefile b/src/backend/Makefile
index b03d5e510f..8c8c53c3b8 100644
--- a/src/backend/Makefile
+++ b/src/backend/Makefile
@@ -37,8 +37,21 @@ LOCALOBJS += utils/probes.o
 endif
 endif
 
+LOCALOBJS += mock.o
+
 OBJS = $(SUBDIROBJS) $(LOCALOBJS) $(top_builddir)/src/port/libpgport_srv.a \
        $(top_builddir)/src/common/libpgcommon_srv.a
+PGLIBOBJS := $(call expand_subsys,$(OBJS))
+PGLIBOBJS += ../pl/plpgsql/src/*.o ../../contrib/pgcrypto/sha1.o
+
+TO_EXCLUDE=commands/% replication/% utils/fmgrtab.o utils/fmgr/fmgr.o utils/adt/rangetypes_typanalyze.o \
+					 access/heap/vacuumlazy.o ./utils/adt/array_typanalyze.o utils/cache/% utils/misc/rls.o \
+					 utils/adt/acl.o utils/misc/superuser.o catalog/aclchk.o catalog/heap.o
+
+PGLIBOBJS := $(filter-out $(TO_EXCLUDE), $(PGLIBOBJS))
+# Readd specific things excluded before
+PGLIBOBJS += commands/define.o
+
 
 # We put libpgport and libpgcommon into OBJS, so remove it from LIBS; also add
 # libldap and ICU
@@ -53,14 +66,22 @@ endif
 
 ##########################################################################
 
-all: submake-libpgport submake-catalog-headers submake-utils-headers postgres $(POSTGRES_IMP)
+all: submake-libpgport submake-catalog-headers submake-utils-headers $(POSTGRES_IMP)
 
 ifneq ($(PORTNAME), cygwin)
 ifneq ($(PORTNAME), win32)
 ifneq ($(PORTNAME), aix)
 
-postgres: $(OBJS)
-	touch $@
+../pl/plpgsql/src/%.o:
+	cd ../pl/plpgsql/; make
+
+pglib.so: $(OBJS)
+	$(CC) $(CFLAGS) $(CPPFLAGS) -Wl,--gc-sections -shared $(call expand_subsys,$^) $(LDFLAGS) $(LDFLAGS_EX) $(export_dynamic) $(LIBS) -o $@
+
+AR_LTO := $(shell dirname `$(CC) -print-libgcc-file-name`)
+
+pglib.a: $(PGLIBOBJS)
+	$(AR) rT $@ --plugin $(AR_LTO)/liblto_plugin.so $(call expand_subsys,$^)
 
 endif
 endif
@@ -71,6 +93,7 @@ ifeq ($(PORTNAME), cygwin)
 postgres: $(OBJS)
 	$(CC) $(CFLAGS) $(call expand_subsys,$^) $(LDFLAGS) $(LDFLAGS_EX) $(export_dynamic) -Wl,--stack,$(WIN32_STACK_RLIMIT) -Wl,--export-all-symbols -Wl,--out-implib=libpostgres.a $(LIBS) -o $@
 
+
 # libpostgres.a is actually built in the preceding rule, but we need this to
 # ensure it's newer than postgres; see notes in src/backend/parser/Makefile
 libpostgres.a: postgres
diff --git a/src/pl/plpgsql/src/Makefile b/src/pl/plpgsql/src/Makefile
--- a/src/pl/plpgsql/src/Makefile
+++ b/src/pl/plpgsql/src/Makefile
@@ -19,8 +19,8 @@
 SHLIB_LINK = $(filter -lintl, $(LIBS))
 rpath =
 
-OBJS = pl_gram.o pl_handler.o pl_comp.o pl_exec.o \
-	pl_funcs.o pl_scanner.o $(WIN32RES)
+OBJS = pl_gram.o pl_handler.o pl_comp.o  \
+	pl_funcs.o pl_scanner.o plmock.o $(WIN32RES)
 
 DATA = plpgsql.control plpgsql--1.0.sql plpgsql--unpackaged--1.0.sql
 

