diff --git a/src/backend/parser/scan.l b/src/backend/parser/scan.l
index 154defc6b8..26a54e1505 100644
--- a/src/backend/parser/scan.l
+++ b/src/backend/parser/scan.l
@@ -420,7 +420,6 @@ numeric_junk	{numeric}{ident_start}
 real_junk		{real}{ident_start}
 
-param			\${decinteger}
+param			\${decdigit}+
-param_junk		\${decinteger}{ident_start}
 
 other			.
 
@@ -1013,10 +1012,6 @@ other			.
 					yylval->ival = atol(yytext + 1);
 					return PARAM;
 				}
-{param_junk}	{
-					SET_YYLLOC();
-					yyerror("trailing junk after parameter");
-				}
 
 {decinteger}	{
 					SET_YYLLOC();
