diff --git a/src/backend/parser/scan.l b/src/backend/parser/scan.l
index 5c930dfda2..0e7b0d488f 100644
--- a/src/backend/parser/scan.l
+++ b/src/backend/parser/scan.l
@@ -421,7 +421,6 @@ real_junk		{real}{ident_start}
 
 /* Positional parameters don't accept underscores. */
 param			\${decdigit}+
-param_junk		\${decdigit}+{ident_start}
 
 other			.
 
@@ -1014,10 +1013,6 @@ other			.
 					yylval->ival = atol(yytext + 1);
 					return PARAM;
 				}
-{param_junk}	{
-					SET_YYLLOC();
-					yyerror("trailing junk after parameter");
-				}
 
 {decinteger}	{
 					SET_YYLLOC();
