--- /Users/taolongxiang/CLionProjects/libpg_query/tmp/postgres/src/include/nodes/primnodes.h	2023-11-07 06:04:27
+++ ./temp/primnodes.h	2024-06-27 16:43:33
@@ -2010,8 +2010,55 @@
 	List	   *fromlist;		/* List of join subtrees */
 	Node	   *quals;			/* qualifiers on join, if any */
 } FromExpr;
+

 /*----------
+ * DistributionType - how to distribute the data
+ *
+ *----------
+ */
+typedef enum DistributionType
+{
+    DISTTYPE_REPLICATION,            /* Replicated */
+    DISTTYPE_HASH,                /* Hash partitioned */
+    DISTTYPE_ROUNDROBIN,            /* Round Robin */
+    DISTTYPE_MODULO,                /* Modulo partitioned */
+    DISTTYPE_SHARD
+} DistributionType;
+
+/*----------
+ * DistributeBy - represents a DISTRIBUTE BY clause in a CREATE TABLE statement
+ *
+ *----------
+ */
+typedef struct DistributeBy
+{
+    NodeTag        type;
+    DistributionType disttype;        /* Distribution type */
+    List        *colname;
+} DistributeBy;
+
+typedef enum PGXCSubClusterType
+{
+    SUBCLUSTER_NONE,
+    SUBCLUSTER_NODE,
+    SUBCLUSTER_GROUP
+} PGXCSubClusterType;
+
+/*----------
+ * PGXCSubCluster - Subcluster on which a table can be created
+ *
+ *----------
+ */
+typedef struct PGXCSubCluster
+{
+    NodeTag                type;
+    PGXCSubClusterType    clustertype;    /* Subcluster type */
+    List                *members;        /* List of nodes or groups */
+} PGXCSubCluster;
+
+
+/*----------
  * OnConflictExpr - represents an ON CONFLICT DO ... expression
  *
  * The optimizer requires a list of inference elements, and optionally a WHERE
